rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,468 @@
+
+### User Impact
+
+The proposed change will address the major pain point brought up by users
+(see [Motivation](#motivation)).  We will announce the feature via a blog post,
+and link to it on all the associated bugs.
+
+There are no special requirements for rolling out the feature.  The Python API
+can be added without changing existing models.
+
+## Detailed Design
+
+To introduce `.at`, we can use a Python `property` to allow it to treat
+`Tensor.at` as a function call.  This function returns a helper object that can
+be used with general indexing to update an underlying tensor:
+```python
+class TensorIndexUpdateHelper:
+  """Helper class that allows copying/modifying a tensor at a set of indices."""
+  def __init__(self, tensor, index_exp):
+    self.tensor = tensor
+    self.index_exp = index_exp
+
+  def set(self, values):
+    """Replaces values in a tensor, returning the modified copy."""
+    # Internal implementation that performs the copy-and-update operation.
+    return tf.internal.copy_and_update(self.tensor, self.index_exp, values)
+
+
+class TensorIndexUpdateHelperFactory:
+  """Creates `TensorIndexUpdateHelper`s via __getitem__."""
+  def __init__(self, tensor):
+    self.tensor = tensor
+
+  def __getitem__(self, index_exp):
+    """Creates a helper via Tensor.at[index_exp]"""
+    return TensorIndexUpdateHelper(self.tensor, index_exp)
+
+
+class Tensor:
+  # ...
+  def create_index_update_helper(self):
+    return TensorIndexUpdateHelperFactory(self)
+
+  # Allows Tensor.at to be treated as a function call, returning a
+  # `TensorIndexUpdateHelperFactory`.
+  at = property(create_index_update_helper)
+```
+
+Through the indirection above, we have
+```python
+#   x.at                     # Returns a TensorIndexUpdateHelperFactory.
+#   x.at[:,:,2]              # Returns a TensorIndexUpdateHelper.
+y = x.at[:,:,2].set(values)  # Sets values in tensor at the specified indices.
+```
+
+For the `tf.internal.copy_and_update` functionality, we will modify the logic
+already in [`array_ops._slice_helper`](https://github.com/tensorflow/tensorflow/blob/3efa4cb01c79e8aa46ca2e80af1a8f85bbc6435d/tensorflow/python/ops/array_ops.py#L915)
+to extract the relevant slice information and call into the appropriate kernels
+to replace values in the slice.  A preliminary version of this is already
+available in the `tf.numpy` package via [`np_array_ops.__slicer_helper`](https://github.com/tensorflow/tensorflow/blob/3efa4cb01c79e8aa46ca2e80af1a8f85bbc6435d/tensorflow/python/ops/numpy_ops/np_array_ops.py#L1509).  The two helpers will be unified and properly tested
+using NumPy's [slice testing](https://github.com/numpy/numpy/blob/c679644b0bc074ca27d9cd26b2aea7cbf4c6f60c/numpy/array_api/tests/test_array_object.py#L25).
+
+## Questions and Discussion Topics
+
+- Does the `.at[]` syntax seem appealing to the TF community?
+- Should we consider deprecating existing gather/scatter methods in favor of
```

Comment:
The existing methods may save users some Python overhead?

Reply:
Good point.  Will this only be a concern for eager?  I don't expect the overhead to be significant, but we can do some basic benchmarking and make sure that the most common use-cases are as fast as possible.

Reply:
It's also a concern for graph-building overhead.

Reply:
I see.  Relative to the rest of graph-building, I again suspect the cost of this will be negligible, but maybe you have found otherwise with the tf-numpy project?  We can do some benchmarking to confirm for models that currently contain `tensor_scatter_nd_update`.

Reply:
We didn't find any real-world problems (but tf-numpy wasn't used much anyway). Agree that benchmarking will help. We can punt on benchmarking until I see real problems.